src/lib/tls/msg_certificate.cpp

```diff
@@ -54,7 +54,35 @@ Certificate::Certificate(const std::vector<byte>& buf)
          throw Decoding_Error("Certificate: Message malformed");
 
       DataSource_Memory cert_buf(&certs[3], cert_size);
-      m_certs.push_back(X509_Certificate(cert_buf));
+      X509_Certificate cert(cert_buf);
+      
+	  std::unique_ptr<Public_Key> cert_pub_key(cert.subject_public_key());
+          
+      const std::string algo_name = cert_pub_key->algo_name();
+      const size_t keylength = cert_pub_key->max_input_bits();
+      if(algo_name == "RSA")
+         {
+         const size_t expected_keylength = policy.minimum_rsa_bits();
+         if(keylength < expected_keylength)
+            throw TLS_Exception(Alert::INSUFFICIENT_SECURITY,
+                                "The peer sent RSA certificate of " +
+                                std::to_string(keylength) +
+                                " bits, policy requires at least " +
+                                std::to_string(expected_keylength));
+         }
+      else if(algo_name == "ECDH")
```

Comment:
This probably depends on the constraints in the certificate (?). If the certificate has wrong constraints and algo name set to ECDH, it can also be used for ECDSA. I think we saw this in our tests.

Probably we should better check for ECDH and ECDSA. Would it be fine?